### PR TITLE
#2859 multiple portfolio flags bug - [KO]

### LINKED ISSUE #2859: Multiple portfolio flags bug: no check when editing an existing portfolio
### Current Behavior

Currently, you cannot add more than one userportfoliopermission per user when the "multiple_portfolios" waffle flag is off - which is expected. However, if you take an existing userportfoliopermission record and add the same user that way, the form lets you do this.

### Expected Behavior

The multiple_portfolios waffle flag is intended to only allow one UserPortfolioPermission record per user.
The current check we have on new records should remain, but we should also deny the user from associating themselves with multiple portfolios by just editing another unrelated UserPortfolioPermission record.

### Steps to Reproduce

1. Create two different UserPortfolioPermission records (with different portfolios and users)
2. On one of the records, set yourself as the user.
3. On the record that isn't set as your user, edit it.
4. Add yourself as the user to that record.
5. Note that now, technically, you are assigned multiple portfolios

### Environment

_No response_

### Additional Context
We're currently doing a form-level check for this, and that should be sufficient for this ticket as well.

Note: this may be related to the self.pk check in user_portfolio_permission.py.
image

### Issue Links

- relates to by #2402

### PR DESCRIPTION
**Pull request by CocoByte**
## Ticket 2859

Resolves #2859

## Changes

- Updated the clean() function in UserPortfolioPermissions to fix the error where users could be added to multiple portfolios via reassignment, even if the multiple portfolio flag is off.
- Added a unit test to check for this error in the future

## Code Review Verification Steps

- [ ] Create two different UserPortfolioPermission records (with different portfolios and users)
- [ ] On one of the records, set yourself as the user.
- [ ] On the record that isn't set as your user, edit it.
- [ ] Add yourself as the user to that record.
- [ ] Note that an error message is shown saying the user cannot be added to multiple portfolios

#### Satisfied acceptance criteria and met development standards

- [ ] Met the acceptance criteria, or will meet them in a subsequent PR
- [ ] Created/modified automated tests
- [ ] Added at least 2 developers as PR reviewers (only 1 will need to approve)
- [ ] Messaged on Slack or in standup to notify the team that a PR is ready for review
- [ ] Changes to “how we do things” are documented in READMEs and or onboarding guide
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the associated migrations file has been commited.

#### Ensured code standards are met (Original Developer)

- [ ] All new functions and methods are commented using plain language
- [ ] Did dependency updates in Pipfile also get changed in requirements.txt?
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values

#### Validated user-facing changes (if applicable)

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)
- [ ] Add at least 1 designer as PR reviewer

#### Reviewed, tested, and left feedback about the changes

- [ ] Pulled this branch locally and tested it
- [ ] Reviewed this code and left comments
- [ ] Checked that all code is adequately covered by tests
- [ ] Made it clear which comments need to be addressed before this work is merged
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the associated migrations file has been commited.

#### Ensured code standards are met (Code reviewer)

- [ ] All new functions and methods are commented using plain language
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values
- [ ] (Rarely needed) Did dependency updates in Pipfile also get changed in requirements.txt?

#### Validated user-facing changes as a developer

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Meets all designs and user flows provided by design/product
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers, the suggestion is to use ones that the developer didn't (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

**Note:** Multiple code reviewers can share the checklists above, a second reviewers should not make a duplicate checklist

#### Verified that the changes match the design intention

- [ ] Checked that the design translated visually
- [ ] Checked behavior
- [ ] Checked different states (empty, one, some, error)
- [ ] Checked for landmarks, page heading structure, and links
- [ ] Tried to break the intended flow

#### Validated user-facing changes as a designer

- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

### CODE REVIEW

File: src/registrar/models/user_portfolio_permission.py [hunk 1]
```diff
@@ -98,10 +97,11 @@ def _get_portfolio_permissions(self):
     def clean(self):
         """Extends clean method to perform additional validation, which can raise errors in django admin."""
         super().clean()
+        TerminalHelper.colorful_logger(logger.info, TerminalColors.OKCYAN, f"**** CLEANING ****")
```
Comment (abroddrick) [inline src/registrar/models/user_portfolio_permission.py:100]: This should be removed, otherwise it will bloat our logs

**Comment by erinysong** [comment]
Verified the following
You can not add a second user portfolio permission  with multiple portfolios flag disabled
image

You can add a second user portfolio permission with multiple portfolios flag enabled
image

File: src/registrar/models/user_portfolio_permission.py [hunk 2]
```diff
@@ -99,20 +99,26 @@ def clean(self):
...
-
         # Check if portfolio is set without accessing the related object.
         has_portfolio = bool(self.portfolio_id)
         if not has_portfolio and self._get_portfolio_permissions():
             raise ValidationError("When portfolio roles or additional permissions are assigned, portfolio is required.")
 
         if has_portfolio and not self._get_portfolio_permissions():
             raise ValidationError("When portfolio is assigned, portfolio roles or additional permissions are required.")
+
+        # Check if a user is set without accessing the related object.
```
Comment (erinysong) [inline src/registrar/models/user_portfolio_permission.py:110]: clarifying question: what does not accessing the related object entail?

File: src/registrar/tests/test_models.py [hunk 3]
```diff
@@ -287,24 +288,53 @@ def test_clean_on_multiple_portfolios_when_flag_active(self):
...
+            cm.exception.message,
+            (
+                "This user is already assigned to a portfolio. "
+                "Based on current waffle flag settings, users cannot be assigned to multiple portfolios."
+            ),
+        )
+
+    @less_console_noise_decorator
+    @override_flag("multiple_portfolios", active=False)
+    def test_multiple_portfolio_reassignment(self):
```
[the remark this answers is not in the document]
Comment (CocoByte) [reply src/registrar/tests/test_models.py:317]: This test specifically sets the multiple_portfolios flag to False before it runs so we shouldn't need to modify  or delete it.

[the remark this answers is not in the document]
Comment (CocoByte) [reply src/registrar/tests/test_models.py:317]: This test specifically sets the multiple_portfolios flag to False before it runs so we shouldn't need to modify  or delete it.

File: src/registrar/models/user_portfolio_permission.py (hunk 2 above)
Comment (CocoByte) [reply src/registrar/models/user_portfolio_permission.py:110]: It means we don't actually check if an object with the given user id exists...we just check that a user id is present.

File: src/registrar/models/user_portfolio_permission.py [hunk 4]
```diff
@@ -99,20 +99,26 @@ def clean(self):
...
+        has_user = bool(self.user_id)
+        if has_user:
+            existing_permission_pks = UserPortfolioPermission.objects.filter(user=self.user).values_list(
+                "pk", flat=True
+            )
+            if (
+                not flag_is_active_for_user(self.user, "multiple_portfolios")
+                and existing_permission_pks.exists()
+                and self.pk not in existing_permission_pks
```
[the remark this answers is not in the document]
Comment (CocoByte) [reply src/registrar/models/user_portfolio_permission.py:119]: No, this check is not to work around fixtures.  It basically makes sure that:
1 - we are not allowing multiple portfolios
2 - the user is already assigned to a portfolio (has an existing portfolio permission)
3 - the portfolio we want to assign to this user is a different one that the already assigned to them

[the remark this answers is not in the document]
Comment (CocoByte) [reply src/registrar/models/user_portfolio_permission.py:119]: But, yes, our fixtures technically allow us to "start out" with multiple portfolios even if the flag is off.  Dealing with that isn't within scope of this ticket.

File: src/registrar/models/user_portfolio_permission.py (hunk 2 above)
Comment (erinysong) [reply src/registrar/models/user_portfolio_permission.py:110]: Thank you for clarifying! Is there a reason we specify here that we check the user id without checking for the object itself? I know it's from existing code but I noticed these comments' wording make it confusing on what they're trying to communicate (ie whether it's relevant that we're looking up the id without the object) so left a suggestion to make the comment more concise. You have more exposure to this model, so feel free to ignore the suggestion if you think it's necessary though!

Comment (erinysong) [inline src/registrar/models/user_portfolio_permission.py:110]:
```suggestion
 # Validate that user is not assigned to another portfolio
```

File: src/registrar/models/user_portfolio_permission.py [hunk 5]
```diff
@@ -99,20 +99,26 @@ def clean(self):
...
-        has_user = bool(self.user_id)
-        if self.pk is None and has_user:
-            existing_permissions = UserPortfolioPermission.objects.filter(user=self.user)
-            if not flag_is_active_for_user(self.user, "multiple_portfolios") and existing_permissions.exists():
-                raise ValidationError(
-                    "This user is already assigned to a portfolio. "
-                    "Based on current waffle flag settings, users cannot be assigned to multiple portfolios."
-                )
-
         # Check if portfolio is set without accessing the related object.
```
Comment (erinysong) [inline src/registrar/models/user_portfolio_permission.py:102]:
```suggestion
Validate required fields for portfolio permission role
```

File: src/registrar/models/user_portfolio_permission.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (erinysong) [reply src/registrar/models/user_portfolio_permission.py:119]: Thank you! I was curious what in the original code wasn't making that check work in the first place

**Review by erinysong — COMMENTED** [review]
Verified ACs and LGTM. Left one nonblocking suggestion

**Review by erinysong — APPROVED** [review]

**Pull request merged**